### ISSUE TITLE
Installation/setup wizard considerations

### ISSUE THREAD

**Issue by rozniak** [rozniak/xfce-winxp-tc#81]
This repo contains many parts that build up the overall 'total conversion'. Themes, icons, cursors, programs, panel plugins etc.

It's a lot of effort, even with instructions, for a user to just jump in and start using it all. What with having to compile the binary portions and run various packaging scripts.

I think the total conversion should cover absolutely everything, including in this case, the setup wizard. The user should be able to kick off setup and have an interface similar to if they were installing XP itself - you know, going with the whole 'look and feel' thing this repo has going for it.

Nothing concrete yet, but I'm making this issue just to sort of brainstorm some ideas on the implementation and wishlist items. In brief, the wizard should obviously deal with compiling and packaging components in this repo, and installing them, perhaps providing the user a checklist akin to 9x's setup on which ones they would like to install.

The interface is my main concern, since XP has both a text-mode and graphical-mode portions of setup.

If someone runs the setup wizard, and X is running, then I believe it should display the Windows XP autorun dialog - prompting them to begin setup. They choose to begin setup and go straight into a graphical-mode portion.

If X is not running, then begin a text-mode setup, using something like the ncurses library for this. The only thing I'm stuck on is the transition from text-mode to graphical mode... do we use text-mode to install X stuff and somehow launch into graphical-mode? I'm not sure how that would work.

And the other thing to consider is the OOBE... I don't know enough about X to really see how we can boot straight into a temporary account or something to do OOBE related setup.

I may update this with additional ideas or proof-of-concept things for UI, any comments welcome. :grin: 

PS: Just writing this down for now as it's in my head, I have no immediate plan to get this working because I think other things take priority but I may do some kind of PoC or comment further if I am in the mood.

**Comment by rozniak** [comment]
Might be worth making a branch for this, and perhaps building up a Dia sheet on the following things:
- Features required (text-mode, graphical-mode, OOBE, autorun... etc.)
- Problems/unknowns (dealing with existing display/window manager that isn't LightDM/XFCE, OOBE etc.)
- Potential solutions (eg. JSON format for optional 9x-style feature list)

Think that it would be easier to build up a document plan rather than commenting mishmash ideas in the issue - better to use this spot for notes perhaps.

**Comment by rozniak** [comment]
Need to consider packaging, particularly updates. Should the installer compile, package and install locally - or should the setup install a custom package repository I host so that updates can be provided that way?

The latter is probably the preferred option between the two...

**Comment by rozniak** [comment]
OOBE most likely will need to be implemented within the LightDM greeter I think? That's what I've been thinking anyway.

Not sure what this entails long-term for Wayland though...

**Comment by rozniak** [comment]
Copied from #250 :

Rough plan for setup:

    Running bootstrapper
        Identify distro, export env vars + check minimum deps to begin (eg. Python)
        No display - go straight to text-mode
        Display present - launch setup autorun for typical 'upgrade' path, copy entire setup tree to disk (because we could be running from an ISO), add setup to init scripts/systemd to launch on next boot into text-mode
    Text-mode setup should install bare minimum deps to be able to start graphical mode setup - add graphical mode setup to init scripts/systemd for reboot
    Graphical mode setup can be a compiled program since at this point the system should have the deps installed, guess this could just be like a minimal env with just X starting with the setup program
        This phase should install remaining components, perhaps give user option for components ala Win9x used to?
        Consider the AUR - see end of comment
        Configuration of XFCE/session complete in this phase
    Chain next boot into OOBE (future work)

Need to consider the AUR, which is another reason I think graphical mode should be compiled/packaged so it can be installed. Should provide some simple shell script that is responsible for chaining phases into the init system which can be run by AUR users to simply skip straight to the configuration portion of setup.

For AUR users it would be something like (setupx is the script here, not necessarily final):

    yay xfce-winxp-tc-git
    setupx --aur
    Reboot
    Should boot into graphical mode setup final phase that just completes the configuration and chains into OOBE

All theoretical, I haven't actually tested much of this and how to handle the various init systems supported. I think it should work but that it is the point of these bootstrap tests.

**Comment by rozniak** [comment]
Begin work after #392 

NTS: Add Artix Linux and Devuan for test against non systemd targets

**Comment by rozniak** [comment]
Looking into this now, think everything is in place to get started on a simple implementation at the very least.

**Comment by rozniak** [comment]
Working on text mode mainly to start with, installing 'base system' packages working:

![Image](https://github.com/user-attachments/assets/7820f7a3-da44-402e-a208-f0792dc22531)

Base system essentially being all libraries, some packages like the native theme, standard cursor theme, graphical setup, and distro packages like lightdm and xfwm4.

Text mode to install that stuff, then prep for reboot: apply plymouth boot screen, add service to init for starting up X session with graphical mode setup

Graphical mode to then install the remaining packages and whatever. Debating how much work might it be to add a 9x style component list... might be something for later though.

Figure at the end of graphical mode, chain into adding service to init for the OOBE.

**Comment by ashdxt** [comment]
working on this, almost done with it, nothing graphical.

currently only works on debian

**Comment by rozniak** [comment]
@ashdxt What part are you working on?

**Comment by ashdxt** [comment]
just a shell script that automates the installation and manual configuration, I want to test it on a few more distros before I upload it, particularly concerned about the things that touch the bootloader.

apologies if this was not the correct channel to communicate this.

**Comment by rozniak** [comment]
Will resume work on GUI mode following the completion of #498 

Another note: Will need to test EndeavourOS with systemd-boot, check Plymouth configuration steps

**Comment by rozniak** [comment]
Sorted out some of the code rot that was in `feat-81`, so setup is 'functional' again on Debian, getting to GUI mode successfully.

Have ported across the dialog templates to GUI mode using the wizard97 lib, everything is generally present now:

<img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/b2f8caf9-3e15-4ff9-888e-7c7785ae0979" />

This is running in test mode on my main install - the actual setup uses Windows Classic via an XFWM4 configuration that gets dropped as part of `startwsetupx` . I have a crappy test for installing `gedit` via setup which works, capturing stdout - so what remains is to pretty much link this up with the package names in `complist.ini` and deploy configs like LightDM and switch systemd over to boot normally.

Aim is to get this up and running on Debian then work on supporting the other targets -- Arch Linux via AUR should be able to skip all the install step and go straight to the config deployment.

**Comment by ashdxt** [comment]
the shell script I made for installation, the installation should work for debian and the steps automating manual configuration should work for everything, I've been thinking of refactoring the code to allow for more customization in which particular settings it implements, let me know if that would be helpful for you.

https://github.com/ashdxt/xfce-winxp-tc-Installer/

the bootloader is experimental, I would need to read up more about grub.

**Comment by rozniak** [comment]
Almost done on this first build of setup, so far I have:
- Text mode launches from setup.sh (no autorun/GUI yet - will error out if DISPLAY or WAYLAND_DISPLAY are set)
- Text mode installs all the WinTC libraries, along with the bits necessary to boot into graphical phase (Plymouth, wintc-setup-gui, fonts etc)
- Text mode adds systemd service to chain into graphical phase after a reboot
- Graphical phase GUI implemented - billboards + wizard steps scaffolded using wizard97
- Graphical phase installs all component packages (they are divided up into groups akin to 9x in the ini, but for now I just install all)
- No OOBE chain yet
- OOBE GUI scaffolded, working playback of intro.wmv and title.wma
- All this packages into an ISO, so user can just mount/extract and run setup.sh to begin 

Remaining stuff:
- OOBE will do the task of applying the end user customisations, for now I plan to have it apply to all interactive users, later on we could have a 'Who will use this computer' step to select specific users
- I think I'll extract the systemd service and boot up script to a separate package to make it more generic between smss and the setup phases instead of each one having it's own copy of the code 
- Testing to ensure this all works reliably

Might push out the resultant ISO as a test for people to try it out, though it will be Debian x64 exclusive. Will open further issues to implement other distros (probably Arch/AUR support as a priority, and Ubuntu/Mint downstreams).

**Comment by rozniak** [comment]
Decided to change how I merge features because squashing all this into 1 commit would be ridiculous I think - the old way was flawed.

---

Finally got this to a point where it works pretty reliably on Debian. Next steps will be to open issues to get things ported over to the other distros - Ubuntu (and downstreams) would be the next closest, and then probably Arch Linux and the rest of the systemd environments. Eventually will deal with FreeBSD and the unusual stuff.

Once I've opened more specific issues I think I will close this one as it has served its purpose.

Created a test release for this work: https://github.com/rozniak/xfce-winxp-tc/releases/tag/rel-547-test
